src/core/vmm.d

```diff
@@ -0,0 +1,489 @@
+else version(Windows)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.decommit | MemoryCapability.reset;
+else version(linux)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.decommit;
+else version(FreeBSD)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.reset;
+else version(OSX)
+    enum MemoryCapability memoryCapability =
+        MemoryCapability.commit | MemoryCapability.reset;
+else
+    static assert(false, "Don't know the memory capability of this OS");
+
+/// The page size as reported by the OS at the application start.
+public alias pageSize = PAGESIZE;
+
+//TODO: provide the same for huge page size
+
+/**
+    Allocate a block of $(D size) bytes of virtual memory rounded up to
+    nearest multiple of page size. The memory is allocated
+    with specified permissions and options.
+
+    $(D base) specifies the desired base address of the block, $(D null)
+    means any base address.
+
+    To release a block of memory allocated by this function use $(LREF free).
+
+    Note: This operation commits virtual memory.
+    To have fine grained control over reserve and commit phases
+    use $(LREF reserve) and $(LREF commit).
+*/
+void[] allocate(void* base, size_t size,
+    MemoryAccess access = MemoryAccess.readWrite,
+    MemoryOptions options = MemoryOptions.none)
+{
+    size = roundUpToPage(size);
+    void* ptr = osMap(base, size, access, options);
+    return ptr ? ptr[0..size] : null;
+}
+
+/**
+    Reserve a block of $(D size) bytes of virtual memory rounded up
+    to nearest multiple of page size. The memory is not commited
+    and/or OS is not explicitly asked to assign memory pages to it.
+
+    $(D base) specifies the desired base address of the block, $(D null)
+    means any base address.
+
+    To release a virtual memory range reserved by this function use $(LREF free).
+
+    Note: This function and $(LREF commit)
+    set read/write permissions but not execute.
+    The reason is that some OS would set permissions on reserve
+    while others on commit. Also Posix makes the commit stage almost transparent.
+    Use $(LREF protect) to override permissions as required.
+*/
+void[] reserve(void* base, size_t size)
+{
+    size = roundUpToPage(size);
+    void* ptr = osReserve(base, size);
+    return ptr ? ptr[0..size] : null;
+}
+
+/**
+    Commit a set of pages that form the specified $(D range) of virtual memory.
+    Only after applying this call is any memory obtained via $(LREF reserve)
+    guaranteed to be accessible.
+
+    Note: This function and $(LREF reserve)
+    set read/write permissions but not execute.
+    The reason is that some OS' would set permissions on reserve
+    while others on commit. Also Posix makes the commit stage almost transparent.
+    Use $(LREF protect) to override permissions as required.
+*/
+bool commit(void[] range)
```

Comment:
I'm slightly concerned that the `void[]` parameter (which will match any mutable array) and the nonspecific names may cause name clashes with other modules. Perhaps the function names could be more explicit, e.g. `commitMemory`? It might also make the user code more readable.